src/Types/TokenName.purs

```diff
@@ -46,17 +53,60 @@ derive newtype instance ToMetadata TokenName
 derive newtype instance Ord TokenName
 derive newtype instance ToData TokenName
 
+asBase16 :: ByteArray -> String
+asBase16 ba = "0x" <> byteArrayToHex ba
+
+fromTokenName :: forall r. (ByteArray -> r) -> (String -> r) -> TokenName -> r
+fromTokenName arrayHandler stringHandler (TokenName cba) = either
+  (\_ -> arrayHandler $ cborBytesToByteArray cba)
+  stringHandler
+  (decodeUtf8 <<< unwrap <<< cborBytesToByteArray $ cba)
+
 instance DecodeAeson TokenName where
+  {-
+      toJSON = JSON.object . Haskell.pure . (,) "unTokenName" . JSON.toJSON .
+        fromTokenName
+            (\bs -> Text.cons '\NUL' (asBase16 bs))
+
+            (\t -> case Text.take 1 t of "\NUL" -> Text.concat ["\NUL\NUL", t]; _ -> t)
+  -}
   decodeAeson = caseAesonObject
     (Left $ TypeMismatch "Expected object")
-    ( note (TypeMismatch "Invalid TokenName") <<< mkTokenName
-        <=< note (TypeMismatch "Invalid ByteArray") <<< hexToByteArray
-        <=< flip getField "unTokenName"
+    ( \aes -> do
+        tkstr :: String <- getField aes "unTokenName"
+        case take 3 tkstr of
+          """\NUL0x""" -> case tkFromStr (drop 3 tkstr) of
+            Nothing -> Left $ TypeMismatch ("Invalid TokenName E1: " <> tkstr)
+            Just tk -> Right tk
+
+          """\NUL\NUL\NUL""" ->
+            note (TypeMismatch $ "Invalid TokenName E2: " <> tkstr)
+              $ tkFromStr (drop 2 tkstr)
+
+          _ -> note (TypeMismatch $ "Invalid TokenName E3: " <> tkstr)
+            $ tkFromStr tkstr
     )
+    where
+    tkFromStr :: String -> Maybe TokenName
+    tkFromStr str = (TokenName <<< wrap) <$>
+      (byteArrayFromIntArray <<< map toCharCode <<< toCharArray $ str)
 
+-- FIXME: what if the tokenname is actually \0\0\0? haskell will break this assuming it
```

Comment:
This should be an issue either on ctl-bridge or CTL

Reply:
This whole thing is a huge PITA! I'll open an issue as soon as I understand how to phrase it :D